File: app/src/main/java/ru/gsench/githubusers/domain/interactor/FavoritesManagement.java
package ru.gsench.githubusers.domain.interactor;

import java.util.ArrayList;

import ru.gsench.githubusers.domain.SystemInterface;
import ru.gsench.githubusers.domain.github_repo.GitHubUserShort;

/**
 * Created by grish on 11.05.2017.
 */

public class FavoritesManagement {

    public static final String FAVORITES = "favorites";
    private static final String divider = "/";

    private ArrayList<UserModel> favorites;
    private SystemInterface system;

    public FavoritesManagement(SystemInterface system){
        this.system=system;
    }

    private void fillFavorites(){
        if(favorites!=null) return;
        String[] fav = system.getSavedStringArray(FAVORITES, new String[0]);
        favorites = new ArrayList<>(fav.length);
        for(String favor: fav){
            String[] div = favor.split(divider);
            favorites.add(new UserModel(
                    Integer.parseInt(div[0]),
                    div[1],
                    true
            ));
        }
    }

    public ArrayList<UserModel> getFavorites(){
        fillFavorites();
        return favorites;
    }

    public boolean isFavorite(GitHubUserShort user){
        fillFavorites();
        for(GitHubUserShort favor: favorites)
            if(favor.getId()==user.getId()) return true;
        return false;
    }

    public ArrayList<UserModel> sortFavorites(ArrayList<GitHubUserShort> users){
        ArrayList<UserModel> ret = new ArrayList<>(users.size());
        for(GitHubUserShort user: users) ret.add(new UserModel(user, isFavorite(user)));
        return ret;
    }

    public void addToFavorites(GitHubUserShort user){
        fillFavorites();
        favorites.add(0, new UserModel(user, true));
        String[] fav = new String[favorites.size()];
        for(int i=0; i<favorites.size(); i++) fav[i]=favorites.get(i).getId()+divider+favorites.get(i).getLogin();
        system.saveStringArray(FAVORITES, fav);
    }

    public void removeFromFavorites(GitHubUserShort user){
        fillFavorites();
        ArrayList<String> fav = new ArrayList<>();
        for(int i=0; i<favorites.size(); i++){
            if(favorites.get(i).getId()!=user.getId())
                fav.add(favorites.get(i).getId()+divider+favorites.get(i).getLogin());
        }
        system.saveStringArray(FAVORITES, fav.toArray(new String[0]));
        for(int i=0; i<favorites.size(); i++) if(favorites.get(i).getId()==user.getId()) favorites.remove(i--);
    }

}

File: app/src/main/java/ru/gsench/githubusers/domain/interactor/FavoritesManagement.kt
package ru.gsench.githubusers.domain.interactor

import java.util.ArrayList

import ru.gsench.githubusers.domain.SystemInterface
import ru.gsench.githubusers.domain.github_repo.GitHubUserShort

/**
 * Created by grish on 11.05.2017.
 */

class FavoritesManagement(private val system: SystemInterface) {

    private var favorites: ArrayList<UserModel>? = null

    private fun fillFavorites() {
        if (favorites != null) return
        val fav = system.getSavedStringArray(FAVORITES, arrayOfNulls<String>(0))
        favorites = ArrayList<UserModel>(fav.size)
        for (favor in fav) {
            val div = favor.split(divider.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            favorites!!.add(UserModel(
                    Integer.parseInt(div[0]),
                    div[1],
                    true
            ))
        }
    }

    fun getFavorites(): ArrayList<UserModel> {
        fillFavorites()
        return favorites
    }

    fun isFavorite(user: GitHubUserShort): Boolean {
        fillFavorites()
        for (favor in favorites!!)
            if (favor.id == user.id) return true
        return false
    }

    fun sortFavorites(users: ArrayList<GitHubUserShort>): ArrayList<UserModel> {
        val ret = ArrayList<UserModel>(users.size)
        for (user in users) ret.add(UserModel(user, isFavorite(user)))
        return ret
    }

    fun addToFavorites(user: GitHubUserShort) {
        fillFavorites()
        favorites!!.add(0, UserModel(user, true))
        val fav = arrayOfNulls<String>(favorites!!.size)
        for (i in favorites!!.indices) fav[i] = favorites!![i].id.toString() + divider + favorites!![i].login
        system.saveStringArray(FAVORITES, fav)
    }

    fun removeFromFavorites(user: GitHubUserShort) {
        fillFavorites()
        val fav = ArrayList<String>()
        for (i in favorites!!.indices) {
            if (favorites!![i].id != user.id)
                fav.add(favorites!![i].id.toString() + divider + favorites!![i].login)
        }
        system.saveStringArray(FAVORITES, fav.toTypedArray())
        var i = 0
        while (i < favorites!!.size) {
            if (favorites!![i].id == user.id) favorites!!.removeAt(i--)
            i++
        }
    }

    companion object {

        val FAVORITES = "favorites"
        private val divider = "/"
    }

}
